compiler/luci/pass/src/FuseAddWithFullyConnectedPass.cpp

```diff
@@ -40,71 +85,26 @@ namespace
  *                |
  *
  */
-bool fuse_add_with_fc(luci::CircleFullyConnected *fc)
+bool fuse_add_with_fc(luci::CircleAdd *add)
 {
-  if (not fc)
-    return false;
+  luci::CircleFullyConnected *fc = nullptr;
+  // addition must be const
+  luci::CircleConst *addition = nullptr;
+  luci::CircleConst *weights = nullptr;
+  luci::CircleNode *bias = nullptr;
 
-  if (fc->dtype() != loco::DataType::FLOAT32)
-    return false;
-
-  if (fc->fusedActivationFunction() != luci::FusedActFunc::NONE)
-    return false;
-
-  auto weights = dynamic_cast<luci::CircleConst *>(fc->weights());
-  if (not weights)
-    return false;
-
-  // Get add node
-  auto fc_output = loco::succs(fc);
-  if (fc_output.size() != 1)
-    return false;
-
-  auto add = dynamic_cast<luci::CircleAdd *>(*fc_output.begin());
-  if (not add)
-    return false;
-  if (add->dtype() != loco::DataType::FLOAT32)
-    return false;
-
-  // Get addition
-  auto addition = add->x() == fc ? dynamic_cast<luci::CircleConst *>(add->y())
-                                 : dynamic_cast<luci::CircleConst *>(add->x());
-
-  // Non-const addition
-  if (not addition)
-    return false;
-
-  auto rank = addition->rank();
-  // TODO Support scalar addition
-  if (rank == 0)
-    return false;
-
-  for (uint32_t i = 0; i < rank - 1; i++)
-  {
-    if (addition->dim(i).value() != 1)
-      return false;
-  }
-  // Check the last dimesion of addition is the same with the number of neurons of FC
-  if (not(addition->dim(rank - 1) == weights->dim(0)))
-    return false;
-
-  auto bias = loco::must_cast<luci::CircleNode *>(fc->bias());
-
-  // We only support (1) constant bias (2) no bias
-  // If bias is neither (1) nor (2), it would be a feature map
-  if (bias->opcode() != luci::CircleOpcode::CIRCLECONST and
-      bias->opcode() != luci::CircleOpcode::CIRCLEOUTPUTEXCLUDE)
-    return false;
+  RETURN_FALSE_UNLESS(
+    common_pass_checks(loco::DataType::FLOAT32, &add, &fc, &addition, &weights, &bias));
 
   auto fused_bias = luci::clone(addition);
 
   // Add existing bias values
   if (auto const_bias = dynamic_cast<luci::CircleConst *>(fc->bias()))
   {
-    assert(const_bias->dtype() == loco::DataType::FLOAT32);
+    RETURN_FALSE_UNLESS(const_bias->dtype() == loco::DataType::FLOAT32);
```

Comment:
Could you check this inside `common_pass_checks`? `luci::clone` adds a node into the graph (graph is modified), so all checks had to be done before `luci::clone`.

Reply:
Ah, this check is specific to float32.. It would be better to use a pattern class (PTAL how `FuseGeluPass` finds different patterns).

Reply:
@jinevening could this refactor be done in separate PR? There is a high chance, I will no longer be at Samsung as of end of this month. So in my opinion, it would be better to push it as-is and create another PR that will introduce pattern search like `FuseGeluPass`.